File: Sources/Vector.swift
#if !os(Linux)
import simd

typealias Vector = double3

extension Vector {
  var lengthp: Double {
    return length(self)
  }

  func dotp(_ v: Vector) -> Double {
    return dot(self, v)
 }

 func crossp(_ v: Vector) -> Vector {
   return cross(self, v)
 }
}
#endif

func ==(lhs: Vector, rhs: Vector) -> Bool {
  return lhs.x == rhs.x && lhs.y == rhs.y && lhs.z == rhs.z
}

extension Vector {
  init(array: [Double]) {
    assert(array.count == 3, "Requires an array of size 3")
    self.init(array[0], array[1], array[2])
  }

  var color: Color {
    return Color(r: Int(x), g: Int(y), b: Int(z))
  }

  var vec3: [Double] {
    return [x, y, z]
  }

  var unit: Vector {
    let l = lengthp
    return Vector(x / l, y / l, z / l)
  }
}

func /(lhs: Vector, rhs: Double) -> Vector {
  return Vector(lhs.x / rhs, lhs.y / rhs, lhs.z / rhs)
}

#if os(Linux)

import Foundation

func +(lhs: Vector, rhs: Vector) -> Vector {
  return Vector(lhs.x + rhs.x, lhs.y + rhs.y, lhs.z + rhs.z)
}

func -(lhs: Vector, rhs: Vector) -> Vector {
  return Vector(lhs.x - rhs.x, lhs.y - rhs.y, lhs.z - rhs.z)
}

func *(lhs: Vector, rhs: Vector) -> Vector {
  return Vector(lhs.x * rhs.x, lhs.y * rhs.y, lhs.z * rhs.z)
}

func *(lhs: Double, rhs: Vector) -> Vector {
  return Vector(lhs * rhs.x, lhs * rhs.y, lhs * rhs.z)
}

func /(lhs: Vector, rhs: Vector) -> Vector {
  return Vector(lhs.x / rhs.x, lhs.y / rhs.y, lhs.z / rhs.z)
}

struct Vector {
  let x: Double
  let y: Double
  let z: Double

  var lengthp: Double {
    return sqrt(x * x + y * y + z * z)
  }

  init(_ x: Double = 0.0, _ y: Double = 0.0, _ z: Double = 0.0) {
    self.x = x
    self.y = y
    self.z = z
  }

  func dotp(_ vec: Vector) -> Double {
    return x * vec.x + y * vec.y + z * vec.z
  }

  func crossp(_ vec: Vector) -> Vector {
    return Vector(
        self.y * vec.z - self.z * vec.y,
        -(self.x * vec.z - self.z * vec.x),
        self.x * vec.y - self.y * vec.x)
  }
}
#endif

File: src-python/vector.py
class Vector:
    def __init__(self, x = 0.0, y = 0.0, z = 0.0):
        self.x = float(x)
        self.y = float(y)
        self.z = float(z)

    @property
    def vec3(self):
        return [self.x, self.y, self.z]

    @property
    def unit(self):
        l = self.lengthp()
        return Vector(
            x = self.x / l,
            y = self.y / l,
            z = self.z / l
        )

    def __add__(self, vec):
        return Vector(self.x + vec.x, self.y + vec.y, self.z + vec.z)

    def __sub__(self, vec):
        return Vector(self.x - vec.x, self.y - vec.y, self.z - vec.z)

    def __mul__(self, vec):
        return Vector(self.x * vec.x, self.y * vec.y, self.z * vec.z)

    def __rmul__(self, s):
        vec = [s * x for x in self.vec3]
        return Vector(vec[0], vec[1], vec[2])

    def __div__(self, vec):
        return self.__truediv__(vec)

    def __truediv__(self, vec):
        return Vector(self.x / vec.x, self.y / vec.y, self.z / vec.z)

    def dotp(self, vec):
        z = zip(self.vec3, vec.vec3)
        r = (x * y for x, y in z)
        return sum(r)

    def crossp(self, vec):
        return Vector(
            x = self.y * vec.z - self.z * vec.y,
            y = -(self.x * vec.z - self.z * vec.x),
            z = self.x * vec.y - self.y * vec.x
        )

    def lengthp(self):
        return sum(x ** 2 for x in self.vec3) ** 0.5
